Permit pydantic v2

ff137 opened the pull request:
Unpin pydantic in the requirements to allow for v2. No breaking changes requiring migration

ff137 commented:
No changes required for migration, hence the PR name: "permit" v2.
But, use of GenericModel [here](https://github.com/permitio/fastapi_websocket_rpc/blob/f8dce437ff788de9856e31fe09550be13be8dc08/fastapi_websocket_rpc/schemas.py#L19) may raise a deprecated warning if the environment is using v2. A fix for that would involve checking the pydantic version, and importing BaseModel instead in case of v2

orweis commented:
Hi @ff137 ! thanks for this great contribution, I'll have someone form the team review soon.
I do think we're already seeing an issue with the tests - perhaps even the one you mentioned.

ff137 commented:
Cool! Thanks.

GenericModel is logged as a deprecation warning, which can be fixed.
But, the error is complaining about `import attr` not found ... So maybe a dependency conflict. I'll take a closer look tomorrow 👌

ff137 commented:
@orweis I've looked again, and the error was `import attr` not found. I can't understand why, because pydantic has no dependencies impacting `attr`. And the only change in this PR is that it now installs pydantic 2.3.0. 

Fortunately, that import is unused! So, I've cleaned up unused imports, and tests pass locally for me. 

I just notice there will indeed be deprecation warnings that can spam logs for users.
The only way that would be to require pydantic >2, and to rename the deprecated methods. 
e.g. "The `json` method is deprecated; use `model_dump_json` instead", and "The `parse_obj` method is deprecated; use `model_validate` instead". 
Trivial to fix those, but for environment compatibility, we'll require pinning pydantic >2. Which should be fine - 2.0 was released more than 2 months ago, so it's about time to switch over :-) and allow people to use `fastapi_websocket_rpc` with the latest pydantic

roekatz commented:
> @orweis I've looked again, and the error was `import attr` not found. I can't understand why, because pydantic has no dependencies impacting `attr`. And the only change in this PR is that it now installs pydantic 2.3.0.
> 
> Fortunately, that import is unused! So, I've cleaned up unused imports, and tests pass locally for me.
> 
> I just notice there will indeed be deprecation warnings that can spam logs for users. One way to address that would be to require pydantic >2, and to rename the deprecated methods. e.g. "The `json` method is deprecated; use `model_dump_json` instead", and "The `parse_obj` method is deprecated; use `model_validate` instead". Trivial to fix those, but for environment compatibility, we'll require pinning pydantic >2. Which should be fine - 2.0 was released more than 2 months ago, so it's about time to switch over :-) and allow people to use `fastapi_websocket_rpc` with the latest pydantic

@ff137 I'm convinced :)
Feel free to make the adjustments you've suggested (pinning pydantic >2, renaming the deprecated methods).

ff137 commented:
> > @orweis I've looked again, and the error was `import attr` not found. I can't understand why, because pydantic has no dependencies impacting `attr`. And the only change in this PR is that it now installs pydantic 2.3.0.
> > Fortunately, that import is unused! So, I've cleaned up unused imports, and tests pass locally for me.
> > I just notice there will indeed be deprecation warnings that can spam logs for users. One way to address that would be to require pydantic >2, and to rename the deprecated methods. e.g. "The `json` method is deprecated; use `model_dump_json` instead", and "The `parse_obj` method is deprecated; use `model_validate` instead". Trivial to fix those, but for environment compatibility, we'll require pinning pydantic >2. Which should be fine - 2.0 was released more than 2 months ago, so it's about time to switch over :-) and allow people to use `fastapi_websocket_rpc` with the latest pydantic
> 
> @ff137 I'm convinced :) Feel free to make the adjustments you've suggested (pinning pydantic >2, renaming the deprecated methods).

Cool! Thank you. I had some ideas for supporting both v1 and v2 of pydantic. Quite inundated with other work atm, but I'll wrap this one up in the next few days

ff137 commented:
So, the deprecated GenericModel usage can be handled quite simply:
```py
import pydantic
from packaging import version
from pydantic import BaseModel

# Check pydantic version to handle deprecated GenericModel
if version.parse(pydantic.VERSION) < version.parse("2.0.0"):
    from pydantic.generics import GenericModel

    class RpcResponse(GenericModel, Generic[ResponseT]):
        result: ResponseT
        result_type: Optional[str]

else:

    class RpcResponse(BaseModel, Generic[ResponseT]):
        result: ResponseT
        result_type: Optional[str]
```

That should be the only adjustment necessary to support pydantic 2 without deprecation warnings.

Note: This requires `packaging` as an extra dependency. It can be avoided by using the built-in `distutils` library, but this is deprecated in python 3.10, so using `packaging` is recommended. It's just for parsing the pydantic version in a reliable way.

ff137 commented:
@orweis @roekatz There we go! I added a way to support both pydantic v1 and v2, without deprecation warnings.
I tested locally with v1 and v2. Pytests pass without any warning logs. Previously it complained about `.json` and `.parse_obj`, which are renamed in v2, and you can see how I implemented it in utils. Let me know if you have any suggestions or change requests

ff137 commented:
A similar approach was followed to add support in the pubsub library: https://github.com/permitio/fastapi_websocket_pubsub/pull/65

roekatz commented on setup.py:
```diff
@@ -13,7 +13,7 @@ def get_requirements(env=""):
 setup(
     name='fastapi_websocket_rpc',
-    version='0.1.24',
+    version='0.1.25',
```
Don't bump the version, we'll do it as part of the release process

ff137 replied:
Hi @roekatz, is the next release in the works? We'd love to use the new pydantic support in production

roekatz replied:
Hi @ff137 , I've actually already released version 0.1.25 of `fastapi_websocket_rpc`.
You probably need a new version of `fastapi_websocket_pubsub` as well? I'll release it now

roekatz replied:
@ff137 Released `fastapi_websocket_pubsub-0.3.8` as well :)
LMK if that works

ff137 replied:
@roekatz aah, I missed it! Because I was checking out the github page. Releases shows:
```
v0.1.23
Latest
on Aug 24, 2022
```
But yes, PyPi is on 0.1.25 :+1:

Similarly the pubsub repo also doesn't have tagged releases synced with the PyPi packages. Not a problem, I guess it just helps with changelogs. 

But yeah, I see pubsub on pypi is still on 0.3.7. The package probably still needs to be published.
Not too urgent for us though -- we still have some more prep work before we can upgrade other things to pydantic 2 as well
So, thanks for following up! I appreciate it

roekatz replied:
@ff137 Oh you're right, accidentally released pubsub 0.3.8 to testpypi instead of pypi.
Now it's released to the pypi as well :)

Thanks for pointing out the missing releases in Github, I'll see if we want to fix that

ff137 replied:
@roekatz Thanks!

roekatz commented on fastapi_websocket_rpc/utils.py:
```diff
@@ -52,9 +57,28 @@ def random_datetime(start=None, end=None):
+        return lambda model, **kwargs: model.model_dump_json(**kwargs)
+
+
+def get_model_parser():
```
Wouldn't it be clearer to have methods `pydantic_parse(model, **kwargs)` and `pydantic_serliaze(model, **kwargs)`.
So one call would be enough instead of first retrieving the correct function and only then using it...

ff137 replied:
Agreed. ChatGPT helped me code that :smile: doing it in one call is better

roekatz reviewed (changes requested):
@ff137 That's great!
Had one request and one suggestion, see my comments. 
Otherwise let's merge that already :)

ff137 commented:
Updated and suggestions applied in _pubsub PR as well :ok_hand:

roekatz commented:
Amazing! :)
